daffodil-runtime1/src/main/scala/org/apache/daffodil/dpath/FNFunctions.scala

```diff
@@ -803,6 +805,74 @@ case class FNFloor(recipe: CompiledDPath, argType: NodeInfo.Kind) extends FNOneA
   }
 }
 
+/**
+ * This is "logical" shift left, meaning it sign-extends a negative signed value.
+ * dfdl:leftShift for signed values does sign-extension, for logical (zero-filling) left shift you must use an unsigned type
+ *
+ * @param recipes
+ * @param argType
+ */
+case class DFDLXLeftShift(recipes: List[CompiledDPath], argType: NodeInfo.Kind) extends FNTwoArgs(recipes)  {
+  override def computeValue(arg1: DataValuePrimitive, arg2: DataValuePrimitive, dstate: DState): DataValuePrimitive = {
+
+    val shiftLong = arg2.getLong
+    val shift = shiftLong.toInt
+    val width = argType.asInstanceOf[PrimNumeric].width.get
+    Assert.invariant(shift >= 0)
+    if(shift>=width)
+      dstate.SDE("Left shift not supported for shift greater or equal to "+ width)
```

Comment:
Should we include the type of the argument and the value of the width passed in to this error message? That might make it more clear that different types have different widths, and what the actual value passed in to the function was. This is helpful if the shift width is calculated, sometimes it's not obvios.

We should also use format strings here. So for example, something like:

```scala
dstate.SDE("Left shift not supported for shift greater or equal to %s for %s. Shift was %s", width, argType.globalQName, shift)
```